scRNA-seq/05-clustering_markers_scRNA.Rmd

```diff
@@ -0,0 +1,349 @@
+This clustering method seeks to find a way to divide the cells into `k` clusters such that the cells within each within each cluster are as similar as possible and the differences among clusters is as large as possible.
+
+It turns out that is a pretty hard problem to solve exactly, but we can do pretty well with an algorithm that starts with a random guess at where the clusters will be:
+
+1. We start by picking random center locations (how we do this can vary)
+2. Then, we assign cells to clusters by finding which center is closest to each cell.
+3. Next we find the centers of these new clusters
+4. Go back to step 2 with these new centers, repeating until the cluster assignments stop changing. 
+
+You might wonder: How many clusters should we use? 
+That is a hard question! 
+There are some heuristics we can use for deciding the "correct" number of clusters, but we will not be exploring those right now. 
+
+For an intuitive visualization of the general k-means method, you might find [this StatsQuest video](https://www.youtube.com/watch?v=4b5d3muPQmA) useful, and for more discussion of the method in a single-cell context, the [Orchestrating Single-Cell Analysis book chapter on k-means](https://bioconductor.org/books/release/OSCA/clustering.html#k-means-clustering) is a good reference.
+
+We are going to use the function `clusterRows()` from the  Bioconductor `bluster` package for our clustering. 
+This function takes a matrix where each sample (cell in our case) is a row and each column is a feature.
+The matrix of counts (or normalized counts) by cells in our `SingleCellExperiment` object is the wrong orientation, so at a minimum we would have to transpose that matrix before proceeding. 
+
+However, clustering algorithms like k-means can be a bit slow with as many features as the number of genes that we have in our data set, so we would rather not use the raw data.
+There is also a potential concern that noise in the raw data might disrupt the clustering algorithm, so it would be best to use some kind of dimensionality reduction algorithm first. 
+We still want to maintain a good number of dimensions, so our old friend PCA is a good (and very standard) choice.
+
+Thankfully, we already computed *and stored* a matrix with reduced dimensions with the `runPCA()` function. 
+We will extract that from the `SingleCellExperiment` object with the `reducedDim()` function, which conveniently returns a matrix with the cells as rows, so we can use that directly!
+
+The other argument we need for `clusterRows()` will tell it which clustering algorithm to use, and any additional parameters associated with that algorithm.
+This has to be made with a special kind of function from the `bluster` package.
+In this case, we will use `KmeansParam()` to specify that we want k-means clustering, with the `centers` parameter to set how many clusters we will assign (`k`).
+
+```{r kmeans_7}
+# set the number of clusters:
```

Comment:
May be a point that you say, but we picked 7 because it "looked good" for instruction purposes, right?

Reply:
Oh, yes, I picked that after trying many things! And b/c random, it only looks good sometimes!